driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java

```diff
@@ -228,187 +185,362 @@ public int getMaxWireVersion() {
+                releaseClientResources();
             }
         }
-    }
 
-    private CommandCursorResult<T> initFromCommandCursorDocument(
-            final ServerAddress serverAddress,
-            final String fieldNameContainingBatch,
-            final BsonDocument commandCursorDocument) {
-        CommandCursorResult<T> cursorResult = new CommandCursorResult<>(serverAddress, fieldNameContainingBatch, commandCursorDocument);
-        cursor.set(cursorResult.getServerCursor());
-        count.addAndGet(cursorResult.getResults().size());
-        postBatchResumeToken = cursorResult.getPostBatchResumeToken();
-        if (LOGGER.isDebugEnabled()) {
-            LOGGER.debug(format("Received batch of %d documents with cursorId %d from server %s", cursorResult.getResults().size(),
-                    cursorResult.getCursorId(), cursorResult.getServerAddress()));
+        void onCorruptedConnection(@Nullable final AsyncConnection corruptedConnection) {
+            // if `pinnedConnection` is corrupted, then we cannot kill `serverCursor` via such a connection
+            AsyncConnection localPinnedConnection = pinnedConnection;
+            if (localPinnedConnection != null) {
+                assertTrue(corruptedConnection == localPinnedConnection);
+                skipReleasingServerResourcesOnClose = true;
+            }
         }
-        return cursorResult;
-    }
 
-    private void logQueryResult(final CommandCursorResult<T> result) {
-        LOGGER.debug(format("Received batch of %d documents with cursorId %d from server %s", result.getResults().size(),
-                result.getCursorId(), result.getServerAddress()));
-    }
+        <R> void executeWithConnection(final Consumer<AsyncConnection> action, final SingleResultCallback<R> callback) {
+            assertTrue(state != State.IDLE);
+            if (pinnedConnection != null) {
+                executeWithConnection(assertNotNull(pinnedConnection).retain(), null, action, callback);
+            } else {
+                assertNotNull(connectionSource).getConnection((conn, t) -> executeWithConnection(conn, t, action, callback));
+            }
+        }
 
-    private class CommandResultSingleResultCallback implements SingleResultCallback<BsonDocument> {
-        private final AsyncConnection connection;
-        private final ServerCursor cursor;
-        private final SingleResultCallback<List<T>> callback;
+        <R> void executeWithConnection(
+                @Nullable final AsyncConnection connection,
+                @Nullable final Throwable t,
+                final Consumer<AsyncConnection> action,
+                final SingleResultCallback<R> callback) {
+            assertTrue(connection != null || t != null);
+            if (t != null) {
+                callback.onResult(null, t);
+            } else {
+                AsyncCallbackSupplier<R> curriedFunction = c -> action.accept(connection);
```

Comment:
The idea behind `AsyncCallbackSupplier` is that at the end, the callback in invoked with the result of the action.  In this case the callback is `c`.  But `c.onResult` is not called here, and because of that, the `whenComplete` action (`connection::release`) does not execute, nor does the callback that's pass to `get` on the next line.    Essentially, this is the equivalent of calling `action.accept(connection` directly.  

So this should be fixed so that it expresses the intention better.  Maybe we just don't need `AsyncCallbackSupplier` at all?  It seems to work properly without it.

Reply:
Thanks - looks like my understanding of AsyncCallbackSupplier was wrong.